### PR TITLE
Multilabel class order

### PR DESCRIPTION
**Pull request by elisno**
Fix class order for multilabel support in Datalab.

The label map previously wasn't always sorted by class names lexicographically.
I've added a property-based test that ensures that internally that the map respects that order (for classification).
I've also added a comment on this on some of the major docs.

Addresses #991

### CODE REVIEW

File: cleanlab/datalab/datalab.py [hunk 1]
```diff
@@ -90,6 +90,9 @@ class Datalab:
     label_name : str, optional
         The name of the label column in the dataset.
 
+        In the case of classification (both multiclass and multilabel), the labels are internally mapped to integers in a [0, 1, ..., K-1] format,
+        ordered lexicographically by class name. Make sure that your model's predictions reflect this ordering.
```
Comment (jwmueller) [inline cleanlab/datalab/datalab.py:94]: this should be for pred_probs. Why does user need to know about this for their labels?

File: docs/source/tutorials/audio.ipynb [hunk 2]
```diff
@@ -478,7 +478,8 @@
...
+    "K-fold cross-validation is a straightforward way to produce out-of-sample predicted probabilities for every datapoint in the dataset, by training K copies of our model on different data subsets and using each copy to predict on the subset of data it did not see during training. An additional benefit of cross-validation is that it provides more reliable evaluation of our model than a single training/validation split. We can obtain cross-validated out-of-sample predicted probabilities from any classifier via the [cross_val_predict](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_val_predict.html) wrapper provided in scikit-learn.\n",
+    "Make sure that the columns of your `pred_probs` are lexicographically ordered by class name.\n"
```
Comment (jwmueller) [inline docs/source/tutorials/audio.ipynb:482]:
```suggestion
    "Make sure that the columns of your `pred_probs` are properly ordered with respect to the ordering of classes, which for Datalab is: lexicographically sorted by class name.\n"
```

Comment (jwmueller) [reply docs/source/tutorials/audio.ipynb:482]: you can apply this feedback everywhere else too

File: docs/source/tutorials/datalab/datalab_quickstart.ipynb [hunk 3]
```diff
@@ -682,6 +683,15 @@
     "examples_w_issue.head()"
    ]
   },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "lab._label_map"
```
Comment (jwmueller) [inline docs/source/tutorials/datalab/datalab_quickstart.ipynb:692]: what is the point of this? we should not expose users to internal class mappings

Comment (elisno) [reply docs/source/tutorials/datalab/datalab_quickstart.ipynb:692]: Sorry, rogue cell.

File: docs/source/tutorials/datalab/datalab_advanced.ipynb [hunk 4]
```diff
@@ -510,7 +510,8 @@
     "To detect certain types of issues in classification data (e.g. label errors), `Datalab` relies on predicted class probabilities from a trained model. Ideally, the prediction for each example should be out-of-sample (to avoid overfitting), coming from a copy of the model that was not trained on this example. \n",
     "\n",
     "This tutorial uses a simple logistic regression model \n",
-    "and the `cross_val_predict()` function from scikit-learn to generate out-of-sample predicted class probabilities for every example in the training set. You can replace this with *any* other classifier model and train it with cross-validation to get out-of-sample predictions."
+    "and the `cross_val_predict()` function from scikit-learn to generate out-of-sample predicted class probabilities for every example in the training set. You can replace this with *any* other classifier model and train it with cross-validation to get out-of-sample predictions.\n",
+    "Make sure that the columns of your `pred_probs` are lexicographically ordered by class name."
```
Comment (elisno) [inline docs/source/tutorials/datalab/datalab_advanced.ipynb:514]:
```suggestion
    "Make sure that the columns of your `pred_probs` are properly ordered with respect to the ordering of classes, which for Datalab is: lexicographically sorted by class name."
```

File: docs/source/tutorials/datalab/datalab_quickstart.ipynb [hunk 5]
```diff
@@ -536,7 +536,8 @@
     "To detect certain types of issues in classification data (e.g. label errors), `Datalab` relies on predicted class probabilities from a trained model. Ideally, the prediction for each example should be out-of-sample (to avoid overfitting), coming from a copy of the model that was not trained on this example. \n",
     "\n",
     "This tutorial uses a simple logistic regression model \n",
-    "and the `cross_val_predict()` function from scikit-learn to generate out-of-sample predicted class probabilities for every example in the training set. You can replace this with *any* other classifier model and train it with cross-validation to get out-of-sample predictions."
+    "and the `cross_val_predict()` function from scikit-learn to generate out-of-sample predicted class probabilities for every example in the training set. You can replace this with *any* other classifier model and train it with cross-validation to get out-of-sample predictions.\n",
+    "Make sure that the columns of your `pred_probs` are lexicographically ordered by class name."
```
Comment (elisno) [inline docs/source/tutorials/datalab/datalab_quickstart.ipynb:540]:
```suggestion
    "Make sure that the columns of your `pred_probs` are properly ordered with respect to the ordering of classes, which for Datalab is: lexicographically sorted by class name."
```

File: docs/source/tutorials/datalab/datalab_quickstart.ipynb [hunk 6]
```diff
@@ -682,6 +683,15 @@
...
   },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "lab._label_map"
+   ]
+  },
```
Comment (elisno) [inline docs/source/tutorials/datalab/datalab_quickstart.ipynb:694]:
```suggestion
```

File: docs/source/tutorials/datalab/tabular.ipynb [hunk 7]
```diff
@@ -220,7 +220,8 @@
    "source": [
     "To find potential labeling errors, cleanlab requires a probabilistic prediction from your model for every datapoint. However, these predictions will be _overfitted_ (and thus unreliable) for examples the model was previously trained on. For the best results, cleanlab should be applied with **out-of-sample** predicted class probabilities, i.e., on examples held out from the model during the training.\n",
     "\n",
-    "K-fold cross-validation is a straightforward way to produce out-of-sample predicted probabilities for every datapoint in the dataset by training K copies of our model on different data subsets and using each copy to predict on the subset of data it did not see during training. We can implement this via the `cross_val_predict` method from scikit-learn:\n"
+    "K-fold cross-validation is a straightforward way to produce out-of-sample predicted probabilities for every datapoint in the dataset by training K copies of our model on different data subsets and using each copy to predict on the subset of data it did not see during training. Make sure that the columns of your `pred_probs` are lexicographically ordered by class name.\n",
```
Comment (elisno) [inline docs/source/tutorials/datalab/tabular.ipynb:223]:
```suggestion
    "K-fold cross-validation is a straightforward way to produce out-of-sample predicted probabilities for every datapoint in the dataset by training K copies of our model on different data subsets and using each copy to predict on the subset of data it did not see during training. Make sure that the columns of your `pred_probs` are properly ordered with respect to the ordering of classes, which for Datalab is: lexicographically sorted by class name.\n",
```

File: docs/source/tutorials/datalab/text.ipynb [hunk 8]
```diff
@@ -258,7 +258,8 @@
     "\n",
     "To identify label issues, cleanlab requires a probabilistic prediction from your model for each datapoint. However these predictions will be _overfit_ (and thus unreliable) for datapoints the model was previously trained on. cleanlab is intended to only be used with **out-of-sample** predicted class probabilities, i.e. on datapoints held-out from the model during the training.\n",
     "\n",
-    "Here we obtain out-of-sample predicted class probabilities for every example in our dataset using a Logistic Regression model with cross-validation."
+    "Here we obtain out-of-sample predicted class probabilities for every example in our dataset using a Logistic Regression model with cross-validation.\n",
+    "Make sure that the columns of your `pred_probs` are lexicographically ordered by class name."
```
Comment (elisno) [inline docs/source/tutorials/datalab/text.ipynb:262]:
```suggestion
    "Make sure that the columns of your `pred_probs` are properly ordered with respect to the ordering of classes, which for Datalab is: lexicographically sorted by class name."
```

File: docs/source/tutorials/image.ipynb [hunk 9]
```diff
@@ -595,7 +595,8 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "Reorder rows of the dataset based on row order in `features` and `pred_probs`. **Carefully ensure your ordering of the dataset matches these objects!**"
+    "Reorder rows of the dataset based on row order in `features` and `pred_probs`. **Carefully ensure your ordering of the dataset matches these objects!**\n",
+    "Make sure that the columns of your `pred_probs` are lexicographically ordered by class name."
```
Comment (elisno) [inline docs/source/tutorials/image.ipynb:599]:
```suggestion
    "Make sure that the columns of your `pred_probs` are properly ordered with respect to the ordering of classes, which for Datalab is: lexicographically sorted by class name."
```

File: docs/source/tutorials/indepth_overview.ipynb [hunk 10]
```diff
@@ -376,7 +376,8 @@
     "All that is need to audit your data is initalize a Datalab object with your dataset and call `find_issues()`. \n",
     "\n",
     "Pass in the predicted probabilities and feature embeddings for your data and Datalab will do all the work!\n",
-    "You do not necessarily need to provide all of this information depending on which types of issues you are interested in, but the more inputs you provide, the more types of issues `Datalab` can detect in your data. Using a better model to produce these inputs will ensure cleanlab more accurately estimates issues."
+    "You do not necessarily need to provide all of this information depending on which types of issues you are interested in, but the more inputs you provide, the more types of issues `Datalab` can detect in your data. Using a better model to produce these inputs will ensure cleanlab more accurately estimates issues.\n",
+    "Make sure that the columns of your `pred_probs` are lexicographically ordered by class name."
```
Comment (elisno) [inline docs/source/tutorials/indepth_overview.ipynb:380]:
```suggestion
    "Make sure that the columns of your `pred_probs` are properly ordered with respect to the ordering of classes, which for Datalab is: lexicographically sorted by class name."
```

File: docs/source/tutorials/multilabel_classification.ipynb [hunk 11]
```diff
@@ -404,7 +404,9 @@
...
-    "Here we produce out-of-sample `pred_probs` by employing cross-validation to fit a multi-label **RandomForestClassifier** model via sklearn's [OneVsRestClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.multiclass.OneVsRestClassifier.html) framework. `OneVsRestClassifier` offers an easy way to apply any multi-class classifier model from sklearn to multi-label classification tasks. It is done for simplicity here, but we advise against this approach as it does not properly model dependencies between classes.\n",
+    "Here we produce out-of-sample `pred_probs` by employing cross-validation to fit a multi-label **RandomForestClassifier** model via sklearn's [OneVsRestClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.multiclass.OneVsRestClassifier.html) framework. \n",
+    "Note that the columns of `pred_probs` need to be lexicographically ordered by class name (i.e. the first column corresponds to the first class, the second column to the second class, and so on).\n",
```
Comment (elisno) [inline docs/source/tutorials/multilabel_classification.ipynb:408]:
```suggestion
    "Make sure that the columns of your `pred_probs` are properly ordered with respect to the ordering of classes, which for Datalab is: lexicographically sorted by class name.\n",
```

File: cleanlab/datalab/datalab.py [hunk 12]
```diff
@@ -191,7 +194,11 @@ def find_issues(
             To best detect label issues, provide this input obtained from the most accurate model you can produce.
 
             For classification data, this must be a 2D array with shape ``(num_examples, K)`` where ``K`` is the number of classes in the dataset.
+                The columns of this array should be lexicographically ordered by class name.
```
Comment (elisno) [inline cleanlab/datalab/datalab.py:197]:
```suggestion
                Make sure that the columns of your `pred_probs` are properly ordered with respect to the ordering of classes, which for Datalab is: lexicographically sorted by class name.
```

File: cleanlab/datalab/datalab.py [hunk 13]
```diff
@@ -191,7 +194,11 @@ def find_issues(
...
             For regression data, this must be a 1D array with shape ``(num_examples,)`` containing the predicted value for each example.
+            For multilabel classification data, this must be a 2D array with shape ``(num_examples, K)`` where ``K`` is the number of classes in the dataset.
+                The columns of this array should be lexicographically ordered by class name.
```
Comment (elisno) [inline cleanlab/datalab/datalab.py:200]:
```suggestion
                Make sure that the columns of your `pred_probs` are properly ordered with respect to the ordering of classes, which for Datalab is: lexicographically sorted by class name.
```

File: docs/source/tutorials/datalab/datalab_quickstart.ipynb (hunk 3 above)
Comment (elisno) [reply docs/source/tutorials/datalab/datalab_quickstart.ipynb]: Removed this cell.

File: cleanlab/datalab/datalab.py (hunk 1 above)
Comment (elisno) [reply cleanlab/datalab/datalab.py:94]: You're right, it's sufficient to just point this out for the pred_probs.

Comment (elisno) [reply cleanlab/datalab/datalab.py:94]: Removed.

**Comment by elisno** [comment]
MacOS instances in the CI are fully queued up across several PRs.

Merging.
